src/agl/WebAppManagerServiceAGL.cpp

```diff
@@ -605,40 +612,55 @@ void WebAppManagerServiceAGL::launchStartupAppFromConfig()
     std::string errMsg;
 
     std::list<ShellSurface *> surfaces;
-    WebAppManagerService::onLaunch(appDesc, params, app_id, errCode, errMsg, surfaces);
+    std::string ret = WebAppManagerService::onLaunch(appDesc, params, app_id, errCode, errMsg, surfaces);
+    delete sargs;
 }
 
-void WebAppManagerServiceAGL::launchStartupAppFromURL()
+void WebAppManagerServiceAGL::launchStartupAppFromURL(void *data)
 {
     LOG_DEBUG("WebAppManagerServiceAGL::launchStartupAppFromURL");
     LOG_DEBUG("    url: %s", startup_app_uri_.c_str());
+
+    WebAppManagerServiceAGL::StartupArgs *sargs =
+        static_cast<WebAppManagerServiceAGL::StartupArgs *>(data);
+
+    if (!sargs)
+        return;
+
+
     Json::Value obj(Json::objectValue);
-    obj["id"] = startup_app_id_;
+    obj["id"] = sargs->GetAppId();
     obj["version"] = "1.0";
     obj["vendor"] = "some vendor";
     obj["type"] = "web";
-    obj["main"] = startup_app_uri_;
+    obj["main"] = sargs->GetAppUri();
     obj["title"] = "webapp";
     obj["uiRevision"] = "2";
     //obj["icon"] = (const char*)icon;
     //obj["folderPath"] = startup_app_.c_str();
-    obj["surfaceId"] = startup_app_surface_id_;
+    obj["surfaceId"] = sargs->GetSurfaceId();
     //obj["surface_role"] = surface_role;
     //obj["panel_type"] = panel_type;
 
-    obj["widthOverride"] = width;
-    obj["heightOverride"] = height;
+    obj["widthOverride"] = sargs->GetWidth();
+    obj["heightOverride"] = sargs->GetHeight();
 
     std::string appDesc;
     dumpJsonToString(obj, appDesc);
-    std::string app_id = startup_app_id_;
+    std::string app_id = sargs->GetAppId();
     int errCode = 0;
     std::string params, errMsg;
 
     LOG_DEBUG("Launching with appDesc=[%s]", appDesc.c_str());
 
     std::list<ShellSurface *> surfaces;
-    WebAppManagerService::onLaunch(appDesc, params, app_id, errCode, errMsg, surfaces);
+    for (AglShellSurface &s : sargs->GetSurfaces()) {
```

Comment:
Adding this here not to forget: Seems here I need here to create these pointers dynamically, and presumably add a virtual dtor then free them later on, after the application has been started. Otherwise I need to access the list directly. GetSurfaces() vs direct access.